### ISSUE TITLE
False-positive InvalidNullableReturnType when non-nullability was previously asserted

### ISSUE THREAD

**Issue by sserbin** [vimeo/psalm#4699]
https://psalm.dev/r/f198dac45f

```php
<?php

function returnsInt(): int {
    $a = rand(0,1) ? 42 : null;
    $b = rand(0,1) ? 84 : null;
    
    assert (!is_null($a) || !is_null($b));
    
    return $a ?? $b;
}
````

expected: no errors
actual: 
```
ERROR: NullableReturnStatement - 9:12 - The declared return type 'int' for returnsInt is not nullable, but the function returns 'int(42)|int(84)|null'

ERROR: InvalidNullableReturnType - 3:24 - The declared return type 'int' for returnsInt is not nullable, but 'int|null' contains null
```
